src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,432 @@
+                  std::optional<LMS_Tree_Node_Idx> leaf_idx,
+                  const LMS_PrivateKey& lms_sk) {
+   auto hash_pair_func = get_hash_pair_func_for_identifier(lms_sk.lms_params(), lms_sk.identifier());
+   auto gen_leaf = lms_gen_leaf_func(lms_sk);
+   TreeAddress lms_tree_address(lms_sk.lms_params().h());
+
+   treehash(out_root,
+            out_auth_path,
+            leaf_idx,
+            lms_sk.lms_params().m(),
+            LMS_TreeLayerIndex(lms_sk.lms_params().h()),
+            0,
+            std::move(hash_pair_func),
+            std::move(gen_leaf),
+            lms_tree_address);
+}
+
+}  // namespace
+
+LMS_Params LMS_Params::create_or_throw(LMS_Algorithm_Type type) {
+   auto [hash_name, height] = [](const LMS_Algorithm_Type& lms_type) -> std::pair<std::string_view, uint8_t> {
+      switch(lms_type) {
+         case LMS_Algorithm_Type::SHA256_M32_H5:
+            return {"SHA-256", 5};
+         case LMS_Algorithm_Type::SHA256_M32_H10:
+            return {"SHA-256", 10};
+         case LMS_Algorithm_Type::SHA256_M32_H15:
+            return {"SHA-256", 15};
+         case LMS_Algorithm_Type::SHA256_M32_H20:
+            return {"SHA-256", 20};
+         case LMS_Algorithm_Type::SHA256_M32_H25:
+            return {"SHA-256", 25};
+         case LMS_Algorithm_Type::SHA256_M24_H5:
+            return {"Truncated(SHA-256,192)", 5};
+         case LMS_Algorithm_Type::SHA256_M24_H10:
+            return {"Truncated(SHA-256,192)", 10};
+         case LMS_Algorithm_Type::SHA256_M24_H15:
+            return {"Truncated(SHA-256,192)", 15};
+         case LMS_Algorithm_Type::SHA256_M24_H20:
+            return {"Truncated(SHA-256,192)", 20};
+         case LMS_Algorithm_Type::SHA256_M24_H25:
+            return {"Truncated(SHA-256,192)", 25};
+         case LMS_Algorithm_Type::SHAKE_M32_H5:
+            return {"SHAKE-256(256)", 5};
+         case LMS_Algorithm_Type::SHAKE_M32_H10:
+            return {"SHAKE-256(256)", 10};
+         case LMS_Algorithm_Type::SHAKE_M32_H15:
+            return {"SHAKE-256(256)", 15};
+         case LMS_Algorithm_Type::SHAKE_M32_H20:
+            return {"SHAKE-256(256)", 20};
+         case LMS_Algorithm_Type::SHAKE_M32_H25:
+            return {"SHAKE-256(256)", 25};
+         case LMS_Algorithm_Type::SHAKE_M24_H5:
+            return {"SHAKE-256(192)", 5};
+         case LMS_Algorithm_Type::SHAKE_M24_H10:
+            return {"SHAKE-256(192)", 10};
+         case LMS_Algorithm_Type::SHAKE_M24_H15:
+            return {"SHAKE-256(192)", 15};
+         case LMS_Algorithm_Type::SHAKE_M24_H20:
+            return {"SHAKE-256(192)", 20};
+         case LMS_Algorithm_Type::SHAKE_M24_H25:
+            return {"SHAKE-256(192)", 25};
+         case LMS_Algorithm_Type::RESERVED:
+            throw Decoding_Error("Unsupported LMS algorithm type");
+      }
+      throw Decoding_Error("Unsupported LMS algorithm type");
```

Comment:
Please use unique error messages: If you see one of those in some bug report log, you'll want to know which code path was taken.

Reply:
... same below, for the hash functions.